[C++] Windows.h Capitalization breaks cross-platform build

### Describe the bug, including details regarding any error messages, version, and platform.

Over at https://github.com/JuliaPackaging/Yggdrasil/pull/9801, we build Arrow cross-platforms for use in the Julia language.

The capitalization of Windows.h in https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/io_util.cc#L121 breaks the mingw compilation for Windows platforms, as the host platform is case-sensitive, and mingw only provides the lowercased variant:

```
[15:22:30] /workspace/srcdir/arrow/cpp/src/arrow/util/io_util.cc:121:12: fatal error: Windows.h: No such file or directory
[15:22:30]  #  include <Windows.h>
[15:22:30]             ^~~~~~~~~~~
[15:22:30] compilation terminated.
```

### Component(s)

C++

--- a/cpp/src/arrow/util/io_util.cc
+++ b/cpp/src/arrow/util/io_util.cc
@@ -118,7 +118,7 @@
 #endif
 
 #ifdef _WIN32
-#  include <Windows.h>
+#  include <windows.h>
 #else
 #  include <dlfcn.h>
 #endif

